### ISSUE TITLE
Error creating widget: could not find model

### ISSUE THREAD

**Issue by SuyashSrivastavaDel** [cdqa-suite/cdQA#248]
Hey @andrelmfarias  and @fmikaelian,
 
I am trying to train reader model on SQuAD 2.0 on GPU with the following code,

      train_processor = BertProcessor(do_lower_case=True, is_training=True, 
                                                               version_2_with_negative=True)
     train_examples, train_features =train_processor.fit_transform(X='./data/SQuAD_2.0/train- v2.0.json')

      reader = BertQA(train_batch_size=12,
                     learning_rate=3e-5,
                     num_train_epochs=2,
                     do_lower_case=True,
                     output_dir='models')
       reader.fit(X=(train_examples, train_features))

After completion of iterations and epochs, I get the following error:
   
      Error creating widget: could not find model

Also, I am planning to only train on GPU but use the model on CPU.
What is incorrect here?
Could you help me with the right direction?

**Comment by andrelmfarias** [comment]
@SuyashSrivastavaDel ,

I have never seen this error before, but by doing a quick search on google I found these issues which mention that the problem might come from jupyterlab:

https://github.com/jupyter-widgets/ipywidgets/issues/2125

https://github.com/bloomberg/bqplot/issues/671

**Comment by SuyashSrivastavaDel** [comment]
Hey @andrelmfarias you were right. Sorry to bother you for this.
I have started to train the model on SQuAD 2.0, will keep you posted for further doubts
Thanks.

**Comment by SuyashSrivastavaDel** [comment]
Hey @andrelmfarias , the training procedure ended, and I am a little confused as to what my next step should be to improve accuracy.

This is the code I ran on JupyterLab with one Tesla V100 GPU as you did:

```
train_processor = BertProcessor(do_lower_case=True, is_training=True, version_2_with_negative=True)
train_examples, train_features = train_processor.fit_transform(X='./data/SQuAD_2.0/train-v2.0.json')

reader = BertQA(train_batch_size=12,
                learning_rate=3e-5,
                num_train_epochs=2,
                do_lower_case=True,
                output_dir='models',
                version_2_with_negative=True)
reader.fit(X=(train_examples, train_features))

reader.model.to('cpu')
reader.device = torch.device('cpu')


joblib.dump(reader, os.path.join(reader.output_dir, 'bert_qa_SQuAD_2_vCPU.joblib'))


```
           

I compared the models on my custom data(11 pdfs(Single Page), each having 3-5 paragraphs), and asked 5 questions related to them, the results were as following:

1. Model trained on SQuAD 2.0 didn't give me an answer for 3 of them, even thought it did give 
    me closest paragraphs and titles related to the question. I understand this happened because 
    thats how  SQuAD 2.0 dataset is made.

2. When compared the given pre-trained(CPU) model with the model I trained, both of them have 
    more or less the same kind of result. They are only able to answer some of the questions 
     correctly. 
    The correctly answered questions weren't same.
3. At this point I am very confused. Should I proceed with fine-tuning one of the trained 
    models on my dataset? If yes, then which model should I use?

4. Or should I train a model on SQuAD 2.0 again, but this time with different hyper-parameters?


Thanks for your time.

**Comment by andrelmfarias** [comment]
Hi @SuyashSrivastavaDel,

No worries!

Your questions are quite challenging actually.

1. From what you say I understand the pipeline find the correct paragraphs but it's not able to give a good answer, am I correct? If it's the case I think the problem comes from the fact that you train the model on SQuAD 2.0. TBH I am afraid it's better to train on SQuAD 1.1 when you know the answers are present in your documents

2. It's normal the correct answers aren't the same, the models are pretty different as they were trained in two different datasets

3 and 4. I think you should try to train the model on SQuAD 1.1 with some different hyperparameters from our official example (also maybe increase the number of epochs?)

I am also currently working on the improvement of the Retriever, maybe it also plays a role in the performance on your specific dataset. Can you try to run only the reader on the paragraphs with the questions about the paragraphs (you can refer to the snippet below for that)? I would like to understand if the poor performance comes from the Reader or the Retriever...

```
import pandas as pd
from sklearn.externals import joblib
from cdqa.utils.converters import df2squad
from cdqa.utils.download import download_model
from cdqa.reader.bertqa_sklearn import BertProcessor, BertQA

download_model(model='bert-squad_1.1', dir='./models')

paragraph = "According to our website, the Melbourne Convention Centre & South Wharf Precinct project is located at 1 Convention Centre Pl, South Wharf VIC 3006, Australia. The Melbourne Convention Centre & South Wharf Precinct project has won three awards. The project started in 2014 and was completed in 2016."
query = "how many awards has the Melbourne Convention Centre project won?"

# Create dataframe and convert it to squad-like json
df = pd.DataFrame({'title': 'Melbourne Convention', 'paragraphs': [[paragraph]]})
json_data = df2squad(df=df, squad_version='v1.1')

# Add question to json
json_data['data'][0]['paragraphs'][0]['qas'].append({"id":0, "question":query})

# Preprocess json
processor = BertProcessor(do_lower_case=True, is_training=False)
examples, features = processor.fit_transform(X=json_data['data'])

# Load model and predict
qa_model = joblib.load("./models/bert_qa_vCPU-sklearn.joblib")

qa_model.predict(X=(examples, features))
```

**Comment by SuyashSrivastavaDel** [comment]
Hey @andrelmfarias thanks for the help. Will check the points above and get back to you.
Also I forgot to mention one thing, while training on SQuAD 2.0 with the code mentioned above,
I landed on this error,

> Training beyond specified 't_total'. Learning rate multiplier set to 0.0. Please set 't_total' of WarmupLinearSchedule correctly.

And then I found [https://github.com/cdqa-suite/cdQA/issues/236#issuecomment-523491362](url)

Changing my code to :
`
reader = BertQA(train_batch_size=12,
                learning_rate=3e-5,
                num_train_epochs=2,
                do_lower_case=True,
                warmup_proportion=0.0,
                output_dir='models',
                version_2_with_negative=True)
reader.fit(X=(train_examples, train_features))
'


I started training a new model(again on SQuAD 2.0) with the code above. Even though the error was thrown again, this time the model I got performed better than the model I have from the code **without** `warmup_proportion=0.0` set to readers parameters. 
I asked the same set of questions on the same dataset and it performed better than the model I trained on SQuAD.

I understand it is for proportion of training to perform linear learning rate warmup and is set to 0.1 initially, but could you help me understand this a little better?

Do I carry along with that error/warning and continue training models?

Thanks.

**Comment by andrelmfarias** [comment]
It's a bit weird, I have never got this warning actually. And I think it's related to hugging face API (the Reader we use is basically their BERT model).
So to be really honest with you, unfortunately, I don't have an answer to your question.

However if ever you figure it out, could you please share it here? I am also curious about it.

**Comment by SuyashSrivastavaDel** [comment]
Hey @andrelmfarias , thanks for replying, I'll get back to you with more updates.
Thanks

**Comment by SuyashSrivastavaDel** [comment]
Hey @andrelmfarias !

Great work on the BM25 retriever.

Experimenting further, I found that in converters, pdf_converter ignores string with length<200 passed to it. In my case, the dataset at some places had single lines pertaining to the same context, for example:
`
There are two types of retriever in cdQA-suite:
1. TF-IDF
2. BM25

`
Now when I ask the question "What are the types of retriever in cdQA-suite?",  the results are not relevant as the converter function skips these lines.

I have made the function configurable in order to make a para out of those lines.
[https://github.com/SuyashSrivastavaDel/cdQA/blob/enhancement/pdf_converter/cdqa/utils/converters.py](url)

Please let me know what you think of this.

Thanks

**Comment by andrelmfarias** [comment]
Hi @SuyashSrivastavaDel thanks for pointing that out!

Could you please open a PR with your changes to the `pdf_converter`? It's better for me to track the changes and make comments. Also, don't forget to merge the lasts changes of the repo to your  fork, I made substantial changes to the master branch these last days...

**Comment by andrelmfarias** [comment]
Oh, also, could you please open another issue for the next time? It doesn't make much sense to having discussions about different subjects in an already closed issue.

Thanks!

**Comment by SuyashSrivastavaDel** [comment]
Hey @andrelmfarias , sure! will keep that in mind!
